[system/pressure] Add plugin to monitor pressure stall information (psi)

HaseHarald opened the pull request:
Adds a plugin to monitor the pressure stall information (psi) as reported by the Linux kernel.

- groups averages per resource
- rate/derive totals for ease of reading
- resources, intervals and scopes configurable

See: https://www.kernel.org/doc/html/latest/accounting/psi.html

I've written this plugin because I read about the metric and wanted to get a feel for what's "normal" on these values. Therefore I did not include any warnings and am not certain that all metrics are actually useful. For example I'm not sure if it helps to monitor the 10 and 60 second averages long term. So I made this stuff configurable. Still this might help someone else, or someone has an idea on an improvement. So here it is for everyone to use.

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+  if [ -d "${pressure_dir}" ]; then
+    printf "yes\n"
+  else
+    printf "no (%s not found)\n" ${pressure_dir}
```
Please add quotes. See the output of `shellcheck` for such issues.

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+}
+
+get_pressure_value() {
+  resource=$1
```
Please use `local` for variables inside of functions in order to keep the namespace clean.

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+  resource=$1
+  interval=$2
+  scope=${3:-some}
+  grep "$scope" ${pressure_dir}//${resource} | grep -o -E "${interval}=[0-9]{1,}(\.[0-9]{1,}){0,1}" | cut -d '=' -f 2
```
Just one slash is sufficient for the `grep` path, oder?

Please quote the path.

Regarding `{1,}` in the regular expression: wouldn't be `+` an easier readable way of expressing this?

HaseHarald replied:
Well, that depends on your definition of "easy". The way it currently is set, the regex works the same way throughout. It's all the same pattern. Whereas your solution (using `+`) brings in a different notation, where one reading this will need to think about it again.

Off course one could argue that this is still common regex that most everybody knows by heart, if they know regex at all, and that regular changes can help to keep you awake. But still its opinion either way around.

The double slash is a typo by the way. I'm going to fix that.

sumpfralle replied:
OK, you thought about the possible ways of expressing this. So I am happy with your choice.

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+      ;;
+  esac
+  
+  printf "%s\n" $printable_name
```
I guess, the linebreak is not relevant here?
(trailing linebreaks are automatically removed in a *Command Substitution*)

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+  done
+}
+
+output_usage() {
```
Maybe simply redirect `output_usage` to *stderr* when it is called?
(in order to remove all these redirects here)

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+}
+
+output_usage() {
+  printf >&2 "%s - munin plugin to graph pressure stall information for CPU, Memory and IO as reported by the Linux kernel.\n" ${0##*/}
```
Maybe suggest to run `munin-doc PLUGIN_NAME` instead of repeating the content of the documentation header here?

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+      config)
+        iterate_config
+        ;;
+      *)
```
`fetch` is also an acceptable alternative for "no argument".

sumpfralle commented on plugins/system/pressure:
```diff
@@ -0,0 +1,331 @@
+  
+  1)
+    case $1 in
+      auto|autoconf)
```
`auto` is not specified. Please treat it as an error.

sumpfralle reviewed (commented):
Thank you for contributing this plugin!

Please take a look at my comments below.

In addition I have the following suggestions:
* please take a look at the issues reported by `shellcheck`
* I would recommend to use a less generic name for this plugin (not just "pressure"). How about `pressure_stall`, `linux_pressure` or `linux_psi`?

Thank you!

HaseHarald commented:
Hello @sumpfralle,

thanks for the feedback. Sorry it took so long to answer, I've been busy. I'm going to go through the issues and resolve them if needed. On some of them I've got a different opinion though. I'll add them right next to your code-quotes so they are easier to relate.

HaseHarald commented:
Hello,

I've fixed several of the issues in 2ee1161a. The once not changed I elaborated above.

Also I renamed the plugin to `linux_psi` which really is the better name.

Have a nice day

HaseHarald

sumpfralle commented:
Your changes look good to me. Thank you!

I merged your plugin now.

btw.: just today another contributor proposed another munin plugin for this topic in [munin#1462](https://github.com/munin-monitoring/munin/pull/1462). Maybe you want to take a look at it.
